### PR TITLE
[None][fix] Self-sampling top-k drops +inf from the top-k (register family)

### PR DESCRIPTION
**Pull request by longcheng-nv**
## Description

Self-sampling GVR V2 decode top-k (CuTe DSL, added in #17821 and made the hint-free production default in #18410) could **drop `+inf` from the top-k** in both register-bracket families:

- `reg` / retained legacy `regimg`: original DKG issue 58 reproduction.
- `reg_clus`: follow-up strict reproduction at `B=4, N=32768, K=1024`.

The failure is independent of hint quality, so it is a selection bug rather than a hint-quality issue.

### Root cause

An in-window `+inf` makes the sampled bracket maximum `+inf`. Then `GMAX - Tv = +inf` and `SC = rcp(+inf) = 0`, collapsing values into a finite histogram bin. The existing sentinel did not rescue this because `SENT_HI - SENT_LO = 6e38` also overflows to `+inf`.

- In `reg` / `regimg`, the collapsed histogram could take the whole-bin emit, so #18501's count-crossing escape did not fire.
- In `reg_clus`, the collapsed histogram could set `whole = 1`, which bypassed the existing exact whole-row fallback.

### Fix

- `GvrTopkRegKernel`: reject an infinite bracket width and force the existing full-row key-space escape when the bracket is invalid.
- `GvrRegClusKernel`: apply the same infinite-width guard, clear `whole`, and mark the result degenerate so rank 0 runs the existing exact whole-row key-space fallback.

Both fallbacks rank `fkey(+inf)` as the maximum key. The finite fast path and the streaming `main` / `clus` kernels are unchanged.

### Correctness contract vs. `torch.topk`

| Input / observable | Current `torch.topk(..., largest=True)` | GVR V2 after this fix | Parity |
|---|---|---|---|
| Finite FP32 values | Selects the numerically largest K values | Tie-aware exact Top-K; indices are valid and unique | Same value multiset |
| `+inf` without NaN | Ranks above every finite value | Retained, including bracket-degenerate cases | Same value multiset |
| Ties, including more than K `+inf` values | May choose any K tied indices; tied indices are not stable | May choose different valid tied indices; output order may differ | Same Top-K semantics, not necessarily identical indices/order |
| NaN | Current CPU/CUDA implementations rank NaN above `+inf` for `largest=True`; tied NaN indices remain unstable | NaN sign/payload and min/max behavior remain implementation-specific | **Not guaranteed to match** |

The comparison contract is the sorted output-value multiset, not element-wise index order. Full PyTorch NaN normalization/selection parity is deferred to future work. The `-inf` duplicate cases fixed by #18501 remain green; the separate BRL `-inf` case tracked by #18502 is outside this PR.

References: [`torch.topk` API and tie behavior](https://docs.pytorch.org/docs/stable/generated/torch.topk.html), [CUDA radix-key NaN handling](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/SortingRadixSelect.cuh).

## Validation (B200, sm100)

- Original issue matrix: in- and out-of-window `+inf`, `N in {4K, 16K, 64K}`, `K in {512, 1024, 2048}`, `cr in {1, 4}` — exact.
- Strict `reg_clus` before/after reproduction (`B=4, N=32768, K=1024`, `+inf` at column 1000): selected `0/1` before, `1/1` in every row after; exact value multiset, valid unique indices, no unwritten slots.
- `reg_clus` stress with `K+5` tied `+inf` values: exactly K unique `+inf` outputs per row.
- GVR V2 family audit: production `main`, `clus`, `reg`, and `reg_clus`, plus retained legacy `regimg`, all pass targeted in-/out-of-sample `+inf` cases. No other GVR V2 omission was found.
- Full `test_gvr_selfsampling_topk.py`: **83 passed**.

### Performance impact

The original `reg` / `regimg` finite-logit A/B remains unchanged:

| Register-family case | Before | Fix | Delta |
|---|---:|---:|---:|
| B1, N4096, K1024 | 6.611 us | 6.593 us | -0.28% |
| B64, N8192, K1024 | 6.644 us | 6.649 us | +0.07% |
| B256, N4096, K1024 | 6.631 us | 6.675 us | +0.66% |
| Pro 4K, N1027/K1024, B256 | 5.721 us | 5.723 us | +0.04% |
| Pro 4K, N1027/K1024, B512 | 5.800 us | 5.802 us | +0.05% |
| Pro 4K, N1027/K1024, B1024 | 5.985 us | 5.919 us | -1.10% |

Same-process paired CUDA-event medians for every routable `reg_clus` template also show no finite-path regression:

| Template | Representative shape | Before | Fix | Delta |
|---|---|---:|---:|---:|
| V1/CS8 | B4, N32768, K1024 | 10.816 us | 10.688 us | -1.18% |
| V2/CS4 | B32, N32768, K1024 | 10.624 us | 10.624 us | 0.00% |
| V4/CS2 | B64, N32768, K1024 | 12.672 us | 12.672 us | 0.00% |
| V2/CS8 | B4, N65536, K1024 | 10.720 us | 10.688 us | -0.30% |
| V4/CS4 | B32, N65536, K1024 | 12.672 us | 12.656 us | -0.13% |
| V4/CS8 | B4, N131072, K1024 | 13.088 us | 13.056 us | -0.24% |

Cubin audit:

- All 9 `reg` / `regimg` varlen variants keep register count and occupancy. The guard adds one `FSETP` and no branch; eight variants keep stack/spills unchanged, while the Pro-4K tuple changes stack 16 -> 24 bytes with four additional `LDL`/`STL`, without a measured regression.
- All 6 routable `reg_clus` templates remain at 64 registers with identical shared/local/stack usage and branch count. The guard adds one `FSETP`; occupancy is unchanged.

A bracket-degenerate `+inf`/`-inf` deliberately takes the correctness fallback. The affected `reg` launch is about 1.5-2.5x slower; the reported `reg_clus` shape is about 5.3x slower because rank 0 performs the whole-row radix descent while the other cluster ranks are idle. Optimizing this exceptional path is future work. Healthy sparse-attention inference is expected to provide finite indexer logits, and the real-capture census found zero NaN/`+inf`/`-inf` among 779,416,662 valid scores, so this PR keeps the fix focused on correctness without changing the normal path.

## Status

- [x] Targeted B200 correctness, performance, and cubin-resource validation completed for both affected kernel families.
- [x] Additional full 886x11 rerun assessed as unnecessary: those captures are finite and the finite fast path is unchanged; targeted exceptional-value tests cover the new behavior.

## PR Checklist

- [x] PR title and description convey what and why
- [x] Test coverage for the change

## Dev Engineer Review

- Updated register-family GVR V2 bracket validation.
- Infinite-width brackets now use the exact key-space fallback.
- The fallback preserves `+inf` by assigning it the maximum key.
- The `reg`, legacy `regimg`, and `reg_clus` families are covered.
- Finite paths and streaming kernels remain unchanged.
- No public API or configuration changes were made.
- B200 validation passed with 83 tests.
- No finite-path performance or kernel resource regression was reported.

## QA Engineer Review

- Added `test_selfsampling_topk_posinf_completeness(pos)`.
- Added `test_selfsampling_topk_posinf_regclus_completeness() -> None`.
- The tests cover register and clustered-register routes.
- The tests verify complete output writes, `+inf` retention, route selection, and exact agreement with `torch.topk`.
- No corresponding entries were identified in `tests/integration/test_lists/`, `test-db/`, or `qa/`.
- Verdict: needs follow-up.

### CODE REVIEW

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/top_k/gvr_topk_decode_self_sampling.py [hunk 1]
```diff
@@ -3713,10 +3713,15 @@ def kern(
...
 
-            # ---- collapse guard, NaN-safe
+            # ---- collapse guard, NaN-safe. An infinite bracket width
+            # (GMAX=+inf from an in-window +inf, or Tv=-inf) makes SC=rcp(w)=0
+            # and folds every value into bin 0, dropping the +inf; reject it
+            # here (okc=0) and force the key-space escape below.
             okc = cutlass.Int32(0)
             if Tv < GMAX:
-                if (GMAX - Tv) > cutlass.Float32(1e-30):
+                w_ = GMAX - Tv
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/top_k/gvr_topk_decode_self_sampling.py:3721]: Thanks. `w_` is defined and consumed entirely within the `if Tv < GMAX` region; only `okc` escapes that region and is therefore predeclared. The current code successfully lowers all nine register-family varlen variants to cubin on B200, and the full 82-test suite passes. No outer predeclaration is required here, so this is a false positive.

**Review by limin2021 — APPROVED** [review]

**Comment by longcheng-nv** [comment]
@limin2021 A follow-up external repro found the same `+inf` collapse in `reg_clus`, so I extended this PR at `9f9203312f`:

- invalid infinite-width brackets now enter the existing rank-0 whole-row key-space fallback;
- added the strict `B=4, N=32768, K=1024` regression (`0/1` `+inf` selected before, exact after);
- full GVR V2 suite: 83 passed; all six `reg_clus` cubins keep 64 registers/occupancy and finite-path A/B shows no regression.

Could you re-review the latest head when convenient?

File: tests/unittest/_torch/thop/parallel/test_gvr_selfsampling_topk.py [hunk 2]
```diff
@@ -292,6 +292,49 @@ def test_selfsampling_topk_neginf_tail_completeness():
...
+    indices = torch.full((1, top_k), -7, dtype=torch.int32, device=_DEV)
+    kv = torch.full((1,), n_valid, dtype=torch.int32, device=_DEV)
+    ss_host.run_varlen(logits, kv, indices, max_seq_len=n_valid)
+    torch.cuda.synchronize()
+    assert int((indices == -7).sum()) == 0, "unwritten output slots"
+    assert torch.isinf(logits[0][indices[0].long()]).any(), "+inf dropped from the top-k"
+    _check_exact(logits, indices, n_valid, ref_vals)
+
+
+def test_selfsampling_topk_posinf_regclus_completeness():
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply tests/unittest/_torch/thop/parallel/test_gvr_selfsampling_topk.py]: Applied in `73af66531c`: the regression test now has an explicit `-> None` return annotation. The targeted test and pre-commit both pass. Thanks.

**Review by zongfeijing — APPROVED** [review]
